File: node/antchain_node/chain.py
"""Blockchain state: validation, the account ledger, difficulty history, and
JSON persistence. Kept intentionally simple for a toy/educational chain:
- Single canonical chain in memory (no side-branch tree); a competing chain
  is only adopted if a peer presents one with strictly higher cumulative
  work, in which case we validate it wholesale and swap it in.
- State (balances, per-sender nonces) is rebuilt by replaying the canonical
  chain from genesis. Fine for chains of the length this project is meant
  to run (thousands of blocks, not millions).
"""
from __future__ import annotations

import json
import threading
from pathlib import Path

from .block import Block, genesis_block
from .consensus import ConsensusParams, MAX_TARGET, quality_target, retarget
from .tsp import generate_instance, reference_length

EPS = 1e-6


class ChainError(ValueError):
    pass


class Blockchain:
    def __init__(self, params: ConsensusParams, data_path: str | None = None):
        self.params = params
        self.data_path = Path(data_path) if data_path else None
        self.lock = threading.RLock()
        self.blocks: list[Block] = [genesis_block()]
        self.balances: dict[str, float] = {}
        self.nonces: dict[str, int] = {}

    # ---------------------------------------------------------------- misc
    @property
    def height(self) -> int:
        return self.blocks[-1].index

    @property
    def tip(self) -> Block:
        return self.blocks[-1]

    # ------------------------------------------------------------- replay
    def compute_targets(self, blocks: list[Block]) -> tuple[list[int], int]:
        """Replays retargeting from genesis over `blocks`. Returns
        (targets_used_for_blocks[1:], target_for_the_block_after_the_last).

        This is the ONLY place difficulty is computed, used identically by
        the live single-block mining path (try_extend, via next_target())
        and the full-chain catch-up path (validate_full_chain). They used to
        be two separate implementations that silently disagreed once a node
        fell behind and needed to validate historical blocks retroactively
        -- validate_full_chain applied retargeting, but next_target() never
        did, so every block after the first retarget boundary looked invalid
        to a node that had to catch up, permanently stalling it.
        """
        targets = []
        t = self.params.initial_target()
        for i in range(1, len(blocks)):
            targets.append(t)
            if i % self.params.retarget_interval == 0:
                start = blocks[i - self.params.retarget_interval]
                end = blocks[i]
                actual = end.timestamp - start.timestamp
                expected = self.params.retarget_interval * self.params.target_block_time
                t = retarget(t, actual, expected)
        return targets, t

    def next_target(self) -> int:
        with self.lock:
            _, t = self.compute_targets(self.blocks)
            return t

    @staticmethod
    def cumulative_work(targets: list[int]) -> int:
        return sum(MAX_TARGET // t for t in targets if t > 0)

    # --------------------------------------------------------- validation
    def validate_block(self, block: Block, prev: Block, target: int,
                        balances: dict[str, float], nonces: dict[str, int]) -> None:
        if block.index != prev.index + 1:
            raise ChainError("bad index")
        if block.prev_hash != prev.hash:
            raise ChainError("bad prev_hash")
        if block.timestamp < prev.timestamp:
            raise ChainError("timestamp not monotonic")
        if block.compute_hash() != block.hash:
            raise ChainError("hash does not match header contents")
        if abs(block.reward - self.params.block_reward) > EPS:
            raise ChainError("bad block reward")
        if len(block.transactions) > self.params.max_txs_per_block:
            raise ChainError("too many transactions")

        instance = generate_instance(block.prev_hash, self.params.n_cities)
        if not instance.is_valid_tour(block.tour):
            raise ChainError("invalid tour: not a permutation of all cities")
        f_x = instance.tour_length(block.tour)
        if abs(f_x - block.f_x) > EPS:
            raise ChainError("declared f(x) does not match recomputed tour length")
        f_ref = reference_length(instance)
        if abs(f_ref - block.f_ref) > EPS:
            raise ChainError("declared f_ref does not match recomputed reference")
        if f_x > self.params.t_opt_frac * f_ref + EPS:
            raise ChainError("solution does not meet quality gate T_opt")

        required_target = quality_target(target, self.params.lam, f_ref, f_x)
        if int(block.hash, 16) >= required_target:
            raise ChainError("hash does not meet quality-weighted target")

        # transactions: signature, sender==address(pubkey), balance, nonce order
        touched_nonces = dict(nonces)
        touched_balances = dict(balances)
        for tx in block.transactions:
            if not tx.verify():
                raise ChainError(f"invalid signature/address on tx {tx.tx_hash}")
            expected_nonce = touched_nonces.get(tx.sender, 0)
            if tx.nonce != expected_nonce:
                raise ChainError(f"bad nonce for {tx.sender}: expected {expected_nonce}, got {tx.nonce}")
            bal = touched_balances.get(tx.sender, 0.0)
            if bal < tx.amount:
                raise ChainError(f"insufficient balance for {tx.sender}")
            touched_balances[tx.sender] = bal - tx.amount
            touched_balances[tx.recipient] = touched_balances.get(tx.recipient, 0.0) + tx.amount
            touched_nonces[tx.sender] = expected_nonce + 1

        touched_balances[block.miner_address] = touched_balances.get(block.miner_address, 0.0) + block.reward
        balances.clear(); balances.update(touched_balances)
        nonces.clear(); nonces.update(touched_nonces)

    def validate_full_chain(self, blocks: list[Block]) -> tuple[dict, dict, list[int]]:
        if not blocks or blocks[0].hash != genesis_block().hash:
            raise ChainError("bad or missing genesis block")
        balances: dict[str, float] = {}
        nonces: dict[str, int] = {}
        targets, _ = self.compute_targets(blocks)
        for i in range(1, len(blocks)):
            self.validate_block(blocks[i], blocks[i - 1], targets[i - 1], balances, nonces)
        return balances, nonces, targets

    # ------------------------------------------------------------ mutate
    def try_extend(self, block: Block) -> bool:
        """Validate `block` as the new tip extending the current chain.
        Returns True and applies it if valid."""
        with self.lock:
            target = self.next_target()
            balances = dict(self.balances)
            nonces = dict(self.nonces)
            self.validate_block(block, self.tip, target, balances, nonces)
            self.blocks.append(block)
            self.balances = balances
            self.nonces = nonces
            self._persist()
            return True

    def try_replace(self, blocks: list[Block]) -> bool:
        """Adopt `blocks` as the new canonical chain if it is valid and has
        strictly more cumulative work than the current chain."""
        with self.lock:
            balances, nonces, targets = self.validate_full_chain(blocks)
            current_targets, _ = self.compute_targets(self.blocks)
            if self.cumulative_work(targets) <= self.cumulative_work(current_targets):
                return False
            self.blocks = blocks
            self.balances = balances
            self.nonces = nonces
            self._persist()
            return True

    def balance_of(self, address: str) -> float:
        with self.lock:
            return self.balances.get(address, 0.0)

    def next_nonce_for(self, address: str) -> int:
        with self.lock:
            return self.nonces.get(address, 0)

    # -------------------------------------------------------- persistence
    def _persist(self) -> None:
        if not self.data_path:
            return
        self.data_path.parent.mkdir(parents=True, exist_ok=True)
        with open(self.data_path, "w") as f:
            json.dump([b.to_dict() for b in self.blocks], f)

    def load(self) -> bool:
        if not self.data_path or not self.data_path.exists():
            return False
        with open(self.data_path) as f:
            raw = json.load(f)
        blocks = [Block.from_dict(d) for d in raw]
        balances, nonces, _ = self.validate_full_chain(blocks)
        self.blocks, self.balances, self.nonces = blocks, balances, nonces
        return True

    def to_dict_list(self) -> list[dict]:
        with self.lock:
            return [b.to_dict() for b in self.blocks]

File: cpp_node/src/chain.cpp
#include "chain.hpp"

#include <cmath>
#include <fstream>
#include <sstream>

#include "tsp.hpp"

namespace antchain_node {

namespace {
constexpr double kEps = 1e-6;
}

Blockchain::Blockchain(ConsensusParams p, std::string data_path)
    : params(std::move(p)), data_path_(std::move(data_path)) {
    blocks.push_back(genesis_block());
}

uint64_t Blockchain::height() const { return blocks.back().index; }
const Block& Blockchain::tip() const { return blocks.back(); }

std::pair<std::vector<U256>, U256> Blockchain::compute_targets(const std::vector<Block>& bs) const {
    std::vector<U256> targets;
    U256 t = params.initial_target();
    for (size_t i = 1; i < bs.size(); ++i) {
        targets.push_back(t);
        if (i % static_cast<size_t>(params.retarget_interval) == 0) {
            const Block& start = bs[i - static_cast<size_t>(params.retarget_interval)];
            const Block& end = bs[i];
            double actual = end.timestamp - start.timestamp;
            double expected = static_cast<double>(params.retarget_interval) * params.target_block_time;
            t = retarget(t, actual, expected);
        }
    }
    return {targets, t};
}

U256 Blockchain::next_target() {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    return compute_targets(blocks).second;
}

U256 Blockchain::cumulative_work(const std::vector<U256>& targets) {
    U256 total;
    for (const auto& t : targets) {
        if (t.is_zero()) continue;
        total = total.add(U256::divide(max_target(), t));
    }
    return total;
}

void Blockchain::validate_block(const Block& block, const Block& prev, const U256& target,
                                 std::map<std::string, double>& bal, std::map<std::string, uint64_t>& nce) const {
    if (block.index != prev.index + 1) throw ChainError("bad index");
    if (block.prev_hash != prev.hash) throw ChainError("bad prev_hash");
    if (block.timestamp < prev.timestamp) throw ChainError("timestamp not monotonic");
    if (block.compute_hash() != block.hash) throw ChainError("hash does not match header contents");
    if (std::abs(block.reward - params.block_reward) > kEps) throw ChainError("bad block reward");
    if (block.transactions.size() > static_cast<size_t>(params.max_txs_per_block))
        throw ChainError("too many transactions");

    TSPInstance instance = generate_instance(seed_from_hash_hex(block.prev_hash), params.n_cities);
    if (!instance.is_valid_tour(block.tour)) throw ChainError("invalid tour: not a permutation of all cities");
    double f_x = instance.tour_length(block.tour);
    if (std::abs(f_x - block.f_x) > kEps) throw ChainError("declared f(x) does not match recomputed tour length");
    double f_ref = reference_length(instance);
    if (std::abs(f_ref - block.f_ref) > kEps) throw ChainError("declared f_ref does not match recomputed reference");
    if (f_x > params.t_opt_frac * f_ref + kEps) throw ChainError("solution does not meet quality gate T_opt");

    U256 required_target = quality_target(target, params.lam, f_ref, f_x);
    U256 block_hash_value = U256::from_hex(block.hash);
    if (!(block_hash_value < required_target)) throw ChainError("hash does not meet quality-weighted target");

    std::map<std::string, uint64_t> touched_nonces = nce;
    std::map<std::string, double> touched_balances = bal;
    for (const auto& tx : block.transactions) {
        if (!tx.verify()) throw ChainError("invalid signature/address on tx " + tx.tx_hash());
        uint64_t expected_nonce = touched_nonces.count(tx.sender) ? touched_nonces[tx.sender] : 0;
        if (tx.nonce != expected_nonce) throw ChainError("bad nonce for " + tx.sender);
        double senderBal = touched_balances.count(tx.sender) ? touched_balances[tx.sender] : 0.0;
        if (senderBal < tx.amount) throw ChainError("insufficient balance for " + tx.sender);
        touched_balances[tx.sender] = senderBal - tx.amount;
        touched_balances[tx.recipient] = (touched_balances.count(tx.recipient) ? touched_balances[tx.recipient] : 0.0) + tx.amount;
        touched_nonces[tx.sender] = expected_nonce + 1;
    }
    touched_balances[block.miner_address] =
        (touched_balances.count(block.miner_address) ? touched_balances[block.miner_address] : 0.0) + block.reward;

    bal = std::move(touched_balances);
    nce = std::move(touched_nonces);
}

std::tuple<std::map<std::string, double>, std::map<std::string, uint64_t>, std::vector<U256>>
Blockchain::validate_full_chain(const std::vector<Block>& bs) const {
    if (bs.empty() || bs.front().hash != genesis_block().hash) throw ChainError("bad or missing genesis block");
    std::map<std::string, double> bal;
    std::map<std::string, uint64_t> nce;
    auto targets_pair = compute_targets(bs);
    const auto& targets = targets_pair.first;
    for (size_t i = 1; i < bs.size(); ++i) {
        validate_block(bs[i], bs[i - 1], targets[i - 1], bal, nce);
    }
    return {bal, nce, targets};
}

bool Blockchain::try_extend(const Block& block) {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    U256 target = next_target();
    std::map<std::string, double> bal = balances;
    std::map<std::string, uint64_t> nce = nonces;
    validate_block(block, tip(), target, bal, nce);
    blocks.push_back(block);
    balances = std::move(bal);
    nonces = std::move(nce);
    persist();
    return true;
}

bool Blockchain::try_replace(const std::vector<Block>& new_blocks) {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    auto [bal, nce, targets] = validate_full_chain(new_blocks);
    auto current_targets = compute_targets(blocks).first;
    if (!(cumulative_work(current_targets) < cumulative_work(targets))) return false;
    blocks = new_blocks;
    balances = std::move(bal);
    nonces = std::move(nce);
    persist();
    return true;
}

double Blockchain::balance_of(const std::string& address) {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    auto it = balances.find(address);
    return it == balances.end() ? 0.0 : it->second;
}

uint64_t Blockchain::next_nonce_for(const std::string& address) {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    auto it = nonces.find(address);
    return it == nonces.end() ? 0 : it->second;
}

void Blockchain::persist() {
    if (data_path_.empty()) return;
    Json arr = Json::array();
    for (const auto& b : blocks) arr.push_back(b.to_json());
    std::ofstream out(data_path_, std::ios::trunc);
    if (!out) throw ChainError("could not open data path for writing: " + data_path_);
    out << arr.dump();
}

bool Blockchain::load() {
    if (data_path_.empty()) return false;
    std::ifstream in(data_path_);
    if (!in) return false;
    std::ostringstream ss;
    ss << in.rdbuf();
    Json arr = Json::parse(ss.str());
    std::vector<Block> loaded;
    for (const auto& bj : arr.as_array()) loaded.push_back(Block::from_json(bj));
    auto [bal, nce, targets] = validate_full_chain(loaded);
    (void)targets;
    blocks = std::move(loaded);
    balances = std::move(bal);
    nonces = std::move(nce);
    return true;
}

Json Blockchain::to_json_list() {
    std::lock_guard<std::recursive_mutex> lock(mutex_);
    Json out = Json::array();
    for (const auto& b : blocks) out.push_back(b.to_json());
    return out;
}

} // namespace antchain_node
